Mathlib/Algebra/Polynomial/Bivariate.lean

```diff
@@ -0,0 +1,96 @@
+/-
+Copyright (c) 2024 Junyan Xu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Junyan Xu
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+
+/-!
+# Bivariate polynomials
+
+This file introduces the notation `R[X][Y]` for the polynomial ring `R[X][X]` in two variables,
+and the notation `Y` for the second variable, in the `Polynomial` scope.
+
+It also defines `Polynomial.evalEval` for the evaluation of a bivariate polynomial at a point
+on the affine plane, which is a ring homomorphism (`Polynomial.evalEvalRingHom`), as well as
+the abbreviation `CC` to view a constant in the base ring `R` as a bivariate polynomial.
+-/
+
+/-- The notation `Y` for `X` in the `Polynomial` scope. -/
+scoped[Polynomial] notation3:max "Y" => Polynomial.X (R := Polynomial _)
+
+/-- The notation `R[X][Y]` for `R[X][X]` in the `Polynomial` scope. -/
+scoped[Polynomial] notation3:max R "[X][Y]" => Polynomial (Polynomial R)
+
+namespace Polynomial
+
+noncomputable section
+
+variable {R S : Type*}
+
+/-- `evalEval x y p` is the evaluation `p(x,y)` of a two-variable polynomial `p : R[X][Y]`. -/
+abbrev evalEval [Semiring R] (x y : R) (p : R[X][Y]) : R := eval x (eval (C y) p)
+
+/-- `evalEval x y` as a ring homomorphism. -/
+@[simps!] abbrev evalEvalRingHom [CommSemiring R] (x y : R) : R[X][Y] →+* R :=
+  (evalRingHom x).comp (evalRingHom <| C y)
+
+/-- A constant viewed as a polynomial in two variables. -/
+abbrev CC [Semiring R] (r : R) : R[X][Y] := C (C r)
+
+lemma coe_algebraMap_eq_CC [CommSemiring R] : algebraMap R R[X][Y] = CC (R := R) := rfl
+lemma coe_evalEvalRingHom [CommSemiring R] (x y : R) : evalEvalRingHom x y = evalEval x y := rfl
+
+lemma evalEvalRingHom_eq [CommSemiring R] (x : R) :
+    evalEvalRingHom x = eval₂RingHom (evalRingHom x) := by
+  ext <;> simp
+
+lemma eval₂_evalRingHom [CommSemiring R] (x y : R) (p : R[X][Y]) :
+    eval₂ (evalRingHom x) y p = evalEval x y p := by
+  rw [← coe_evalEvalRingHom, evalEvalRingHom_eq]; rfl
```

Comment:
Wouldn't this be more natural of a direction to state?
```
example (x y : R) (p : R[X][Y]) : p.evalEval x y = (p.map <| evalRingHom x).eval y
```
This says evaluating `p` at `y` then at `x` is the same as evaluating `p` at `x` then at `y`.

Reply:
Maybe this could be `evalEval_comm` or `evalEval_eq`, or `map_evalRingHom_eval` if it's reversed.

Reply:
I guess there's no harm in having both. You can replace the `rfl` with `coe_eval₂RingHom` too.

Reply:
I added your version, which just differ by a `Polynomial.eval_map`. I consider `evalEval` to be the simp normal form so I put it on the RHS.

Reply:
Thanks!